### ISSUE TITLE
TouchMove intermediate points are not DPI aware

### ISSUE THREAD

**Issue by dotMorten** [dotnet/wpf#2072]
* .NET Core Version: 3.1 Preview1
* Windows version: (`winver`) 1809
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes
* Is this bug related specifically to tooling in Visual Studio (e.g. XAML Designer, Code editing, etc...)? No

 **Problem description:**
Using the TouchMove event in a multi-dpi scenario, the interpolated points returned are not correctly adjusting for DPI, whereas the current point is.
When using the event to draw a path, you can see the discrepancy between these two inputs:
![image](https://user-images.githubusercontent.com/1378165/67240439-571cd800-f406-11e9-8b49-6baaad9a1a52.png)
In this case my main monitor is scale factor 1.5x, whereas my touch-monitor where I'm running the app on is 1.0x.
The app has been configured to support multi-monitor DPI via the app.manifest.

 
 **Actual behavior:** 
Touch events (red line) are about 30% off from the correct location. If I multiple the location with 1.5, the location is correct (apart from the distance of the window top-bar). This is however weird, as the screen the app is running on is 1.0x screen scale (but main monitor is 1.5x). I'd guess the app registers it needs to run at 1.5 screen scale, then realizes the app is now running at 1.x and but forgot intermediate points should no longer be divided by 1.5x.

 **Expected behavior:**
 Intermediate and current touch points coincide.

 **Minimal repro:**
Xaml:
```xml
    <Grid>
        <Canvas Margin="30" x:Name="canvas" TouchMove="canvas_TouchMove" IsManipulationEnabled="True" Background="White" />
    </Grid>
```
xaml.cs code-behind:
```cs
     private PathFigure intermediatePoints;
        private PathFigure currentPoint;

        private void canvas_TouchMove(object sender, TouchEventArgs e)
        {
            foreach (var p in e.GetIntermediateTouchPoints(canvas))
                intermediatePoints = AddVertex(p.Position, intermediatePoints, Colors.Red, 3);
            currentPoint = AddVertex(e.GetTouchPoint(canvas).Position, currentPoint, Colors.Blue, 1);
        }

        private PathFigure AddVertex(Point point, PathFigure pf, Color color, double width)
        {
            if (pf == null)
            {
                pf = new PathFigure(point, new PathSegment[] { }, false);
                PathGeometry pg = new PathGeometry(new PathFigure[] { pf });
                Path p = new Path() { Data = pg };
                p.StrokeThickness = width;
                p.Stroke = new SolidColorBrush(color);
                canvas.Children.Add(p);
            }
            else
            {
                pf.Segments.Add(new System.Windows.Media.LineSegment(point, true));
            }
            return pf;
        }
```
app.manifest:
```xml
<?xml version="1.0" encoding="utf-8"?>
<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
  <assemblyIdentity version="1.0.0.0" name="MyApplication.app"/>
  <trustInfo xmlns="urn:schemas-microsoft-com:asm.v2">
    <security>
      <requestedPrivileges xmlns="urn:schemas-microsoft-com:asm.v3">
        <requestedExecutionLevel level="asInvoker" uiAccess="false" />
      </requestedPrivileges>
    </security>
  </trustInfo>

  <compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1">
    <application>
      <!-- Windows 10 -->
      <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}" />
    </application>
  </compatibility>

  <application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings>
      <dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">PerMonitor</dpiAwareness>
      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</dpiAware>
    </windowsSettings>
  </application>
</assembly>
```
Make sure you include the manifest in csproj:
```
  <PropertyGroup>
    <ApplicationManifest>app.manifest</ApplicationManifest>
  </PropertyGroup>
```
Repro proejct:
[TouchMoveDpiRepro.zip](https://github.com/dotnet/wpf/files/3752777/TouchMoveDpiRepro.zip)

**Comment by rladuca** [comment]
WPF stores its own set of intermediate points for touch moves due to the way in which it gets touch data in the first place.  Since touch data comes off the message loop via a shared memory channel with WISP, there used to be an issue with flooding the UI thread with touch moves.  We implemented a coalescing scheme in order to deal with that.

That's done here:
https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispLogic.cs#L162

This is just the raw data directly from WISP though.  The actual transformation and return is handled here:
https://github.com/dotnet/wpf/blob/8c7ab1592b6c70e732106aeabf58f215d4b71c22/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Common/StylusTouchDeviceBase.cs#L95

I'll have to look further as to why the data isn't transforming correctly, but those are the investigative starting points.

**Comment by dotMorten** [comment]
Yeah I'm poking around too, and found it curious that the intermediate touch points does some transformation here:
https://github.com/dotnet/wpf/blob/024e02575f753bf58fad556d130501a213658b62/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Common/StylusTouchDeviceBase.cs#L104
 where's it doesn't do that for the current point:
https://github.com/dotnet/wpf/blob/024e02575f753bf58fad556d130501a213658b62/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Common/StylusTouchDeviceBase.cs#L46

**Comment by rladuca** [comment]
Both GetPosition and GetStylusPoints do transformations.  GetPosition is a lot less complicated as it doesn't reformat the collection.

See here:
https://github.com/dotnet/wpf/blob/024e02575f753bf58fad556d130501a213658b62/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs#L1151

The reformatting is likely the culprit here:

https://github.com/dotnet/wpf/blob/ae1790531c3b993b56eba8b1f0dd395a3ed7de75/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Common/StylusPointCollection.cs#L477

**Comment by dotMorten** [comment]
Added the GetPosition to the GetIntermediateTouchPoints method, and the difference between the values returned are pretty clear (see pinned values):

![image](https://user-images.githubusercontent.com/1378165/67244648-065dad00-f40f-11e9-9d87-7e2ec2b4cf4f.png)

**Comment by dotMorten** [comment]
The reformat method just moves the pixels 30 pixels up and over:
![image](https://user-images.githubusercontent.com/1378165/67244880-8f74e400-f40f-11e9-8383-33611c0dce9b.png)

**Comment by dotMorten** [comment]
Looking at GetPosition and stepping through that, it's a completely different way to transform:
https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs#L1220-L1222

![image](https://user-images.githubusercontent.com/1378165/67245059-f5616b80-f40f-11e9-848a-afb52798c4ee.png)

**Comment by rladuca** [comment]
If your window is very close to the top left of the screen, how far off are the numbers?

WISP's raw data is relative to the client area IIRC.

**Comment by dotMorten** [comment]
It doesn't look like it's relative to the screen, but it is indeed relative to the window.
![dpioffset](https://user-images.githubusercontent.com/1378165/67245650-5473b000-f411-11e9-8d07-f53d8b1b2e10.gif)

**Comment by rladuca** [comment]
I don't have a touch screen right now (not in the office) and I don't have time to write injection code just now, but I can take a look a deeper look this week assuming you don't root cause it before then.

Just poking around, this set of code makes me curious as to how we are storing and returning the raw data:
https://github.com/dotnet/wpf/blob/8c7ab1592b6c70e732106aeabf58f215d4b71c22/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs#L1455

I think we're missing some transformation to this for returning the intermediate points.

**Comment by dotMorten** [comment]
OK. I'll poke around more as I'm set up to build at test out fixes, but feel free to give more pointers as this part of WPF is completely unknown to me.

**Comment by dotMorten** [comment]
So this is most likely not the correct fix (I wouldn't be surprised if it broke something else as I don't fully understand this code), but I managed to hack a fix into `WispStylusDevice`'s `UpdateState(RawStylusInputReport)` method:

![image](https://user-images.githubusercontent.com/1378165/67251183-0798d500-f423-11e9-9e30-76eb164882da.png)

**Comment by rladuca** [comment]
@dotMorten Awesome!  It seemed suspect that we weren't applying that transformation coming out of the intermediate points as well.

I think we don't want to alter the actual raw points that are stored here.  We probably just want to add to the transform coming out of `GetIntermediateTouchPoints`.

**Comment by dotMorten** [comment]
This bit in WispLogic is also suspect: Why is the transform 1.5, when on this specific screen it's only 1.0? 
![image](https://user-images.githubusercontent.com/1378165/67251856-53e51480-f425-11e9-81e9-b9c625f0fbd4.png)

**Comment by dotMorten** [comment]
Actually that comment about caching led me into something else. This shouldn't be cached permanently (but it is) when you enable multi-dpi support. The app starts up on my 1.5x screen, and I then move it to the 1.0 screen. However after first getting initialized to 1.5x, it never gets updated to 1.0x.

https://github.com/dotnet/wpf/blob/c951bb6f3ded8971600b47b1f91ab63634f13739/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispLogic.cs#L3113-L3121

**Comment by dotMorten** [comment]
I tried a different setup where the main monitor was a scale 1.0x, and the touch monitor was 1.75x (ie Surface Laptop connected to external display, and the external display set as the main monitor). Same problem appears where the app launches at main monitor, and you move the display to the laptop and screen scale is wrong, but in this case in the "opposite" direction). This is probably a more typical setup and so should be easy to reproduce.

**Comment by rladuca** [comment]
Refreshing the transform on DpiChanged (and probably DpiChangedAfterParent) should work if that's the key part of the change.

@vatsan-madhavan do you think we would need both events here?

We may need to deal with handling this conversion mid-input stream.  The real time thread will be queuing moves or what not at input priority as the window transitions.  Doesn't necessarily mean much as the primary pointer will be controlling the window if it is dragged across, so it's possible secondary pointers get some strange changes since the input is not synchronous with the message loop (and therefore not synchronous with DpiChanged events).  Just something to think a bit about and test out, but that seems like a pretty low percentage scenario.

In a system-aware application, I believe the current code is fine as the message pump will be sending messages in system-aware coordinates and the thread DPI awareness should always be system as well.  So this seems like it is a regression from Per-Mon DPI, probably since .NET Framework 4.6.2.

**Comment by dotMorten** [comment]
> So this seems like it is a regression from Per-Mon DPI, probably since .NET Framework 4.6.2.

Agreed. Well you couldn't even do per-mon DPI before 4.6.2 if I recall, so might have been a non-issue then.

**Comment by rladuca** [comment]
@dotMorten Yep, that's what I meant.

I appreciate you getting to the root of this, by the way.  I spent a lot of time on the touch and stylus stack a couple of years back so I like to see things improving in that area.

**Comment by dotMorten** [comment]
>  I spent a lot of time on the touch

And I spent a lot of time trying to smooth my touch drawing, and finding I need to use TouchMove's intermediate points, only to find that I can't actually use it (unless you have some hints that I can detect this case and just disable the optimization for those few people who actually have this setup). So I'm pretty anxious to get it fixed, so at least I can have the fix on our .NET Core build :-)

**Comment by dotMorten** [comment]
So here's a different way to fix it in `WispLogic.cs` (probably shouldn't be listening to the event like this, but proof-of-concept fix at least):

![image](https://user-images.githubusercontent.com/1378165/67259994-c23dcd00-f44d-11e9-85ea-20a85420837d.png)

If you have some guidance how to do this "the right way" I'd be more than happy to make the PR.

**Comment by rladuca** [comment]
I think it likely the transform needs to be stored per HwndSource.  It's probably a better idea, overall, to store it in each PenContext and then use the appropriate transform corresponding to the latest input.

The problem with storing it in WispLogic is that each HwndSource, in a per-mon scenario, would potentially need a different transform depending on the monitor it's on.  PenContexts are created per tablet/stylus per HwndSource.  So those seem like the perfect place to put it.  Of course the logic surrounding usage of the transform would have to be modified to use the transform from the PenContext appropriate to the input (which is available in the `RawStylusInputReport`.

https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Common/RawStylusInputReport.cs#L110

There might be some other cleanup surrounding this depending on all the paths that hit the transform, but there should be a way of cleanly returning the correct value.  In a system-aware scenario you'd just be caching the same transform everywhere.  So it's slightly less efficient, but at least it's correct for per-mon scenarios.

**Comment by dotMorten** [comment]
@rladuca I'm a little confused about this. WispLogic is where the transform is currently handled, and it's tied to the specific hwnd. I'm not sure how you could have a single hwnd with different transforms at the same time.

It sort of sounds like you're trying to also solve a different broader issue with a much larger and riskier refactoring of how all of this is currently done. Or am I misunderstanding?

**Comment by rladuca** [comment]
@dotMorten Let's say you have two HwndSources in a per-mon DPI aware application.  One is on a monitor with 100% DPI, one is on a monitor with 150% DPI.  The transformation would be different for each of them.  Right now, WispLogic is setup for system-aware DPI.  In that case you only ever need a single transform, the transform corresponding to the primary monitor.  But if you just re-used that transform via a DPI change event, you'd be ping-ponging between different DPI transforms and get it wrong a lot of the time.

So the issue here is where to cache and when.  Updating the cached transform when you get a DPI change is definitely the right way to go.  However, WispLogic itself probably isn't the best place to store this anymore.  Since we're already instantiating a PenContext per tablet for each HwndSource (and these are tied to the input that comes from them) we can store the cached transform there.  It's even fine for the PenContext to subscribe to the DPI changed events since we tie that to the HwndSource's lifetime in almost all cases (barring a tablet refresh).

**Comment by dotMorten** [comment]
@rladuca Took a little stab at that, but it quickly becomes a little bit beyond what I'm comfortable refactoring without a lot of guidance / pair programming. Seems like a significant set of changes because most of the existing methods that rely on the `_transformToDevice` matrix don't have a reference to the pen context, so the ripple effects of the change seems rather large (or I'm going down the wrong road trying to address this).

**Comment by weltkante** [comment]
On a side note, I think the `_transformToDevice` needs to be removed (or store an identity when you are in per-display mode), otherwise this probably will cause problems depending on which monitor you start out.

**Comment by rladuca** [comment]
@dotMorten @weltkante I agree, WispLogic has some pretty ingrained assumptions about DPI support.  

@dotMorten I was looking briefly at getting some sort of workaround for this.  I don't think there is any straightforward way to wind back the intermediate point transforms so you can correctly apply the updated HwndSource transform.  You'd have to rely on a whole lot of implementation details that would be potentially fragile.

I'll take a rough stab at this when I have some time this week and see what I can come up with.  I doubt I will have a full-fledged solution here, but I might be able to give you a branch with some more direction.

**Comment by dotMorten** [comment]
So for my workaround is to detect if the process is multi-dpi-aware, and if so ignore the intermediate points. Unfortunately that means the touch points doesn't give as good a curve as possible, but it will at least be better for all the apps that doesn't enable this in the manifest (which I think is most).

**Comment by rladuca** [comment]
I took a look at this, and I think the changes are going to require some re-thinking of a few pieces of code in the WISP stack.

I was going to store the cached matrix in the `PenContexts` class that has a mapping in `WispLogic` (`PresentationSource` to `PenContexts`).  The problem here is that the transformations are sometimes used to re-acquire a new `InputSource` if the old one became invalid during the time the input took to go from the StylusThread to the UI thread.

https://github.com/dotnet/wpf/blob/8c7ab1592b6c70e732106aeabf58f215d4b71c22/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs#L1526

So that got me thinking about per-mon DPI and how we are processing points prior to actually having the window.  I think that code above might need changing to something more akin to:

https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Window.cs#L7595

Modifying that paradigm to deal with getting coordinates from a `PenContext` and trying to find them in the correct screen (much like we're doing for saved Top/Left coords).  I'll have to put more thought into how to do this correctly.

Anyway, I think this just belies a broader set of issues with the WISP based stack not working well in Per-Monitor DPI modes.

The best way to deal with the basic storing of the matrix might be to (in `WispLogic`) just store it per `DpiScale2` we see and then look that up based on the `DpiScale2` of the `HwndSource` we are dealing with.  This should work fine since `TransformToDevice` is just a scaling based on DPI.  

https://github.com/dotnet/wpf/blob/52f24ad1517fc4412c989ce684189d7249abd137/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/InterOp/HwndTarget.cs#L2324

If we don't find a cached Matrix for that `DpiScale2`, then we query the actual `TransformToDevice` matrix.  I think that's probably the best way to cache it as the various functions seemingly always have access to an `HwndTarget` via `WispStylusDevice.CriticalActiveSource`.  We can just internally expose the `DpiScale2` stored there for our purposes.

I'll try to dedicate some more time to this soon, just to keep tracing through the potential touch points (awful term to use in this context) of all the code changes we'd need.

**Comment by rladuca** [comment]
Working on changes in this branch: https://github.com/rladuca/wpf/tree/permonstylus.

Feel free to pick up and go from what I have (or alter it if you find something wrong).

I am going to do some very basic testing at some point today, but as of the posting of this comment I just verified it builds.

**Comment by rladuca** [comment]
@dotMorten I took your repro and tested it out.

On my main touch screen (200% scale) the lines tracked correctly.  I then moved it to my secondary touch screen (125% scale) and they still tracked.

You can see the results in the screenshot.  THe "M" is from the main screen, the "S" is the secondary.

![image](https://user-images.githubusercontent.com/20006987/67804617-8ce83f00-fa4c-11e9-9a90-338a4bfa136a.png)

There is more testing and evaluation to be done here, but the set of changes seems promising.

**Comment by rladuca** [comment]
@dotMorten Modified your repro to open multiple windows and that all seemed to work as well.

After some thought, I don't really see the point in not caching this in `HwndTarget` itself.  It's already responding to DPI changes to update the DPI scale, so it can just update the transform as well when that happens.  I'll think about that further, but that's a lot nicer than caching it in WispLogic.

**Comment by weltkante** [comment]
I don't know how defensive you want to be, but you probably want to be careful that you aren't handing out direct references to a cached mutable object through public API. Users could edit it and mess up the cache, its probably a good idea to always hand out a frozen instance or a copy, or only make the cached instance available through internal API.

**Comment by TanayBhalani** [comment]
I have followed below article, and found it is now working perfect for me.
I am using .NET Framework 4.5.2

[PerMonitorDPI](https://github.com/Microsoft/WPF-Samples/tree/master/PerMonitorDPI)

**Comment by vatsan-madhavan** [comment]
> Can you see any reason to not just cache the following transform and update on DPI changes?
> 
> https://github.com/dotnet/wpf/blob/52f24ad1517fc4412c989ce684189d7249abd137/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/InterOp/HwndTarget.cs#L2324

It's not meant to be cached normally. Unless you are trying to keep track of the `TransformToDevice` at the time you decide to cache it (i.e, a point-in-time value, or a historical value), just keep track of the associated `HwndTarget` directly and look up the value of the transform whenever its needed - that way the transform would always be current.

**Comment by rladuca** [comment]
> It's not meant to be cached normally. Unless you are trying to keep track of the `TransformToDevice` at the time you decide to cache it (i.e, a point-in-time value, or a historical value), just keep track of the associated `HwndTarget` directly and look up the value of the transform whenever its needed - that way the transform would always be current.

I mean to finish this off soon, the fix I have in my branch sidesteps needing the HwndTarget transform.  Since these transforms solely depend on the DpiScale2, I'm just mapping that to a cached transform.

**Comment by rladuca** [comment]
@fabiant3 Bumping this to triage and unassigning it.  The PR is waiting for review.

**Comment by walterlv** [comment]
![stylus-move-points-not-pm-dpi](https://user-images.githubusercontent.com/9959623/85247631-fbb55500-b480-11ea-9396-810d3c153adf.gif)

I've also tested the per-monitor DPI and find the same issue.

**Comment by ryalanms** [comment]
@rladuca's fix is in master (.NET 6.0) and has been approved for the 5.0.4 servicing release.
